### ISSUE TITLE
Pass enriched metadata to upcaster protocol

### ISSUE THREAD

**Issue by slashdotdash** [commanded/commanded#339]
The `Commanded.Event.Upcaster` protocol allows events to be upcast by providing the source event and its metadata.  It would be useful to have access to the additional fields provided by the `Commanded.EventStore.RecordedEvent` struct, such as the event's `created_at` timestamp. 

Currently event metadata enrichment is only [applied to event handlers](https://github.com/commanded/commanded/blob/v1.0.0/lib/commanded/event/handler.ex#L681). Instead, this enrichment could be moved before upcasting to make it available to any event processors.

### Example usage

```elixir
defimpl Commanded.Event.Upcaster, for: AnEvent do
  def upcast(%AnEvent{} = event, metadata) do
    %AnEvent{name: name} = event
    %{:created_at => created_at} = metadata
    
    IO.inspect(metadata)
    # %{
    #   :causation_id => "db1ebd30-7d3c-40f7-87cd-12cd9966df32",
    #   :correlation_id => "1599630b-9c38-433c-9548-0dd793108ba0",
    #   :created_at => #DateTime<2017-10-30 11:19:56.178901Z>,
    #   :event_id => "5e4a0f38-385b-4d57-823b-a1bcf705b7bb",
    #   :event_number => 12345,
    #   :stream_id => "e42a588d-2cda-4314-a471-5d008cce01fc",
    #   :stream_version => 1,
    #   "issuer_id" => "0768d69a-d2b7-48f4-d0e9-083a97f7ebe0",
    #   "user_id" => "user@example.com"
    # }

    %AnEvent{event | first_name: name, created_at: created_at}
  end
end
```

**Comment by vasspilka** [comment]
Hey, can I open a PR for this?

**Comment by slashdotdash** [comment]
@vasspilka Yes, please do.

**Comment by vasspilka** [comment]
@slashdotdash PR oppened

**Comment by slashdotdash** [comment]
Fixed by #389.
